src/Test/TestLogger.php

```diff
@@ -0,0 +1,328 @@
+<?php
+
+namespace Psr\Log\Test;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\LoggerTrait;
+
+/**
+ * Used for testing purposes.
+ *
+ * It records all records and gives you access to them for verification.
+ */
+class TestLogger implements LoggerInterface
+{
+    use LoggerTrait;
+
+    public array $records = [];
+    public array $recordsByLevel = [];
+
+    /**
+     * @inheritdoc
+     */
+    public function log($level, string|\Stringable $message, array $context = []): void
+    {
+        $record = [
+            'level' => $level,
+            'message' => $message,
+            'context' => $context,
+        ];
+
+        $this->recordsByLevel[$record['level']][] = $record;
+        $this->records[] = $record;
+    }
+
+    /**
+     * @param string $level
+     * @return bool
+     */
+    public function hasRecords($level)
+    {
+        return isset($this->recordsByLevel[$level]);
+    }
+
+    /**
+     * @param array $record
+     * @param string $level
+     * @return bool
+     */
+    public function hasRecord($record, $level)
+    {
+        if (is_string($record)) {
+            $record = ['message' => $record];
+        }
+
+        return $this->hasRecordThatPasses(function ($rec) use ($record) {
+            if ($rec['message'] !== $record['message']) {
+                return false;
+            }
+            if (isset($record['context']) && $rec['context'] !== $record['context']) {
+                return false;
+            }
+            return true;
+        }, $level);
+    }
+
+    /**
+     * @param string $message
+     * @param string $level
+     * @return bool
+     */
+    public function hasRecordThatContains($message, $level)
+    {
+        return $this->hasRecordThatPasses(fn ($rec) => str_contains($rec['message'], $message), $level);
+    }
+
+    /**
+     * @param string $regex
+     * @param string $level
+     * @return bool
+     */
+    public function hasRecordThatMatches($regex, $level)
+    {
+        return $this->hasRecordThatPasses(fn ($rec) => preg_match($regex, $rec['message']) > 0, $level);
+    }
+
+    /**
+     * @param callable $predicate
+     * @param string $level
+     * @return bool
+     */
+    public function hasRecordThatPasses(callable $predicate, $level)
+    {
+        if (!isset($this->recordsByLevel[$level])) {
+            return false;
+        }
+        foreach ($this->recordsByLevel[$level] as $i => $rec) {
+            if ($predicate($rec, $i)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @deprecated Since psr/log-util 1.1
+     */
+    public function __call($method, $args)
```

Comment:
Do we need this deprecation? This is a new class in a new namespace and a new package.

Reply:
This is a new package, but this class existed in psr/log 1.1. 
https://github.com/php-fig/log/blob/1.1.4/Psr/Log/Test/TestLogger.php

If the method is removed, it would exist with `fig/log-test 1.0.*` but not `fig/log-test ^1.1`.

Perhaps there is a decision to change the namespace. I'm not aware of it.

Reply:
This PR is about 1.1 for 2.0/3.0 - We should not have a name collision as we use a different namespace though.

Reply:
I see. 👍🏻